### PR TITLE
add rand_seed. rand_bytes reuse buffer b

### PR DESCRIPTION
**Pull request by aaronUniontech**
Signed-off-by: zhangya <zhangya@uniontech.com>

### CODE REVIEW

File: src/tpm2-tss-engine-rand.c [hunk 1]
```diff
@@ -67,12 +108,13 @@ rand_bytes(unsigned char *buf, int num)
         memcpy(buf, &b->buffer, b->size);
         num -= b->size;
         buf += b->size;
-        free(b);
     }
 
-    esys_ctx_free(&esys_ctx);
-
  end:
+    if(b)
```
Comment (williamcroberts) [inline src/tpm2-tss-engine-rand.c:114]: free is skipped if b is null, no need to check.

File: src/tpm2-tss-engine-rand.c [hunk 2]
```diff
@@ -40,6 +40,47 @@
...
+{
+    ESYS_CONTEXT *esys_ctx = NULL;
+    TSS2_RC r;
+
+    r = esys_ctx_init(&esys_ctx);
+    ERRchktss(rand_seed, r, goto end);
+
+    TPM2B_SENSITIVE_DATA stir;
+    size_t offset = 0;
+    char* cur_data = (char*)seed;
```
Comment (williamcroberts) [inline src/tpm2-tss-engine-rand.c:58]: pointer asterisk tight to variable name not type.
`char *cur_data`

File: src/tpm2-tss-engine-rand.c [hunk 3]
```diff
@@ -40,6 +40,47 @@
...
+
+    static const size_t tpm_random_stir_max_size = 128; 
+    while(offset < (size_t)seed_len) {
+        size_t left = seed_len - offset;
+        // in test, tpm stir seed size beyond 128 will failed with 0x000001d5
```
Comment (williamcroberts) [inline src/tpm2-tss-engine-rand.c:63]: This is in the spec part 3 commands for TPM2_StirRandom, so link to it or mention it here over "it fails, not sure why".

**Review by williamcroberts — CHANGES_REQUESTED** [review]

File: src/tpm2-tss-engine-rand.c [hunk 4]
```diff
@@ -40,6 +40,46 @@
...
+        stir.size = chunk;
+        memcpy(stir.buffer, cur_data + offset, chunk);
+
+        r = Esys_StirRandom(
+            esys_ctx,
+            ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+            &stir);
+        ERRchktss(rand_seed, r, goto end);
+
+        offset += seed_len;
```
Comment (AndreasFuchsSIT) [inline src/tpm2-tss-engine-rand.c:74]: Shouldn't this read `offset += chunk` ?

File: src/tpm2-tss-engine-rand.c [hunk 5]
```diff
@@ -67,12 +107,12 @@ rand_bytes(unsigned char *buf, int num)
         memcpy(buf, &b->buffer, b->size);
         num -= b->size;
         buf += b->size;
-        free(b);
```
Comment (AndreasFuchsSIT) [inline src/tpm2-tss-engine-rand.c:70]: This change is incorrect. Esys_GetRandom allocates a new buffer on every invocation.

**Review by AndreasFuchsSIT — CHANGES_REQUESTED** [review]
See my comments please.

File: src/tpm2-tss-engine-rand.c (hunk 4 above)
Comment (aaronUniontech) [reply src/tpm2-tss-engine-rand.c]: yes. it's serious.

File: src/tpm2-tss-engine-rand.c (hunk 5 above)
Comment (aaronUniontech) [reply src/tpm2-tss-engine-rand.c:70]: the code in tpm2-tss, master, it's only new buffer while it's null. 

    Esys_GetRandom_Finish

/* Allocate memory for response parameters */
    if (randomBytes != NULL) {
        *randomBytes = calloc(sizeof(TPM2B_DIGEST), 1);
        if (*randomBytes == NULL) {
            return_error(TSS2_ESYS_RC_MEMORY, "Out of memory");
        }
    }

**Comment by AndreasFuchsSIT** [comment]
So all changes to `rand_bytes()` should not be part of this PR, but be a different PR, since they are unrelated to the purpose of StirRandom.
Also you need to squash all your commits into a single one, that includes some more information and also a signed-off tag.
Then, since this is a new functionality, I'd like to see a second commit, that introduces a test script invoking the new functionality.

One we got this, I'll be very happy to merge this.
Sorry for so many requests at once, but that's what makes a good PR. :-)

**Comment by aaronUniontech** [comment]
> 

ok

**Pull request closed without merging**